### ISSUE TITLE
Support calendars for parsing text

### ISSUE THREAD

**Issue by mj1856** [nodatime/nodatime#384]
I was looking at a few StackOverflow questions about parsing Persian calendar input strings, and I thought for sure that we could handle that in Noda Time - but I guess not.

Given a string like `"1391/2/30"` - shouldn't I be able to get a `LocalDate` somehow by using the pattern API?  I would think I could call `CalendarSystem.GetPersianCalendar()` and pass the result into `LocalDatePattern.Create`, but we have only overloads that take `CultureInfo`, not `CalendarSystem`.  

And alas, the expected culture `"fa-IR"` does not have the BCL `PersianCalendar` as its default.

Searching the web, there are a lot of questions around this, as the normal idea of passing the culture in to `DateTime.Parse` doesn't work either.  Surely Noda Time can make this problem simpler?

**Comment by jskeet** [comment]
Set the template value to one in the calendar you want. More docs for this definitely required :)

**Comment by mj1856** [comment]
Hmmm.. Both `LocalDatePattern.TemplateValue` and `LocalDatePattern.TemplateValue.Calendar` are read only properties.

**Comment by mj1856** [comment]
Oh, nevermind, I now see the overload in the `LocalDatePattern` constructor that takes a template value.   Though that is highly unintuitive...

**Comment by mj1856** [comment]
Here's what I came up with:

```c#
using System.Globalization;
using NodaTime;
using NodaTime.Text;

var calendar = CalendarSystem.GetPersianCalendar();
var culture = CultureInfo.CreateSpecificCulture("fa-IR");
var pattern = LocalDatePattern.Create("yyyy/M/d", culture, new LocalDate(1, 1, 1, calendar));
var result = pattern.Parse("1391/2/30");
```

**Comment by mj1856** [comment]
This works, but I still think we could make it more intuitive with overloads of `Create` that take a `CalendarSystem`.  Perhaps consider for 2.0.  Thanks.

**Comment by jskeet** [comment]
I think doing it as an overload would be tricky, but we could have a "WithCalendar" method to return a new one with the template value converted into the given calendar system.

It's important that this is feasible, and also that it's easier to discover - but I don't mind it being a couple of operations rather than one, given that I expect it to be reasonably rare.

**Comment by malcolmr** [comment]
@jskeet so something like the following?

```c#
var pattern = LocalDatePattern.Create("yyyy/M/d", culture).WithCalendar(calendar);
```
It seems a little odd that we'd handle culture and calendar differently, since they both seem fairly essential to parsing the string. On the other hand, I guess we've made a fairly strong stance for Iso-as-default calendar, but still.

**Comment by jskeet** [comment]
Yup, that's roughly what I'd expect. Given that it will update the template value, I think it's reasonable for it to be a separate method call - and as you say, we've maintained the position that using multiple cultures is common, but using non-Gregorian calendars isn't. I'm still comfortable with that. Note that you don't need any of this for *formatting*, as there the value knows the calendar already. Oh, and you can already parse a value if contains a calendar ID, although that would be pretty specialist.

**Comment by jskeet** [comment]
Looking back on Malcolm's comment, I've worked out why it's reasonable to handle them differently - the calendar system is part of the result value, which is why it *is* in the template. Compare that with the culture, which is lost as soon as we've finished parsing, effectively.

But yes, we should still make it easier to handle :)

**Comment by jskeet** [comment]
Fixed in commit 366e541.

**Comment by mj1856** [comment]
Thanks!
